### PR TITLE
Change input predicate offset to message & coin

### LINKED ISSUE #166: Include message predicate metadata into the transaction
Currently, only coin predicates are tracked in the transaction metadata. However, we also need to support message predicate metadata.

We could consider consolidating the predicate metadata between coin and message input types, since the data that needs to be tracked is the same for both input types.

### PR DESCRIPTION
**Pull request by vlopes11**
Resolves #166

### CODE REVIEW

File: tests/bytes.rs [hunk 1]
```diff
@@ -858,10 +858,10 @@ fn script_input_coin_data_offset() {
                         );
 
                         let (offset, len) = tx
-                            .input_coin_predicate_offset(offset)
+                            .input_predicate_offset(offset)
```
Comment (Voxelot) [inline tests/bytes.rs:861]: we should update this test to include both coin and message predicates

**Comment by Voxelot** [comment]
do we still need `Input::coin_predicate_len`? Seems like we only need `Input::predicate_len` now.

**Comment by vlopes11** [comment]
> do we still need `Input::coin_predicate_len`? Seems like we only need `Input::predicate_len` now.

True, should be either deprecated or removed. Since we are already breaking the API, I just removed them:

https://github.com/FuelLabs/fuel-tx/pull/174/commits/e8faf6ca42a12dbbb984dbb41671b170b4eef0eb

File: tests/bytes.rs [hunk 2]
```diff
@@ -728,11 +745,25 @@ fn create_input_coin_data_offset() {
...
+                        &buffer[offset..offset + len][..predicate.len()]
+                    );
+
+                    let (offset, len) = tx
+                        .input_predicate_offset(input_message_idx)
                         .expect("Failed to fetch offset");
 
                     let (offset_p, _) = tx_p
-                        .input_coin_predicate_offset(last_input)
+                        .input_predicate_offset(input_message_idx)
```
Comment (Voxelot) [inline tests/bytes.rs:766]: I just realized this test is named `create_input_coin_data_offset`. We should probably rename it or split the message offsets into their own tests.

File: tests/bytes.rs [hunk 3]
```diff
@@ -858,10 +889,10 @@ fn script_input_coin_data_offset() {
                         );
 
                         let (offset, len) = tx
-                            .input_coin_predicate_offset(offset)
+                            .input_predicate_offset(offset)
                             .expect("Failed to fetch offset");
 
-                        assert_ne!(predicate.len(), len);
+                        assert_ne!(bytes::padded_len(&predicate), predicate.len());
```
Comment (Voxelot) [inline tests/bytes.rs:895]: Do we need to test message predicate offsets for scripts too? Looks like it was only added for the create tx offset test.

File: tests/bytes.rs (hunk 2 above)
Comment (vlopes11) [reply tests/bytes.rs:766]: Wouldn't be strictly false to have that as generic: https://github.com/FuelLabs/fuel-tx/pull/174/commits/7ea93d5c0711493c2fa459d04b2009785f785ae6

File: tests/bytes.rs (hunk 3 above)
Comment (vlopes11) [reply tests/bytes.rs:895]: The offset is not bound to the transaction variant. We can ofc test that as sanity check, but in practice it won't increase the robustness of the tests

File: src/builder.rs [hunk 4]
```diff
@@ -12,12 +12,28 @@ use alloc::vec::Vec;
...
+    fn with_tx(tx: Transaction) -> Self {
+        let should_prepare_script = false;
+        let should_prepare_predicate = false;
+        let sign_keys = Vec::new();
+
+        Self {
+            tx,
+            should_prepare_script,
+            should_prepare_predicate,
+            sign_keys,
```
Comment (Dentosal) [inline src/builder.rs:33]: Nit: consider more concise initialization without let bindings

**Review by Voxelot — APPROVED** [review]

**Pull request merged**